src/evaluators/bc/PHAL_SDirichletField_Def.hpp

```diff
@@ -197,45 +198,80 @@ evaluateFields(typename Traits::EvalData dirichlet_workset)
+    if (fill_residual)
+    {
+      printf("Fill residual...\n");
+      auto tf = Albany::getTpetraVector(f);
+      fView = tf->getLocalViewDevice(Tpetra::Access::ReadWrite);
+    }
 
-  auto f_view = fill_residual ? Albany::getNonconstLocalData(f) : Teuchos::null;
-  Teuchos::Array<ST> entries;
-  Teuchos::Array<LO> indices;
-  Teuchos::Array<ST> value(1);
-  value[0] = dirichlet_workset.j_coeff;;
-
-  this->set_row_and_col_is_dbc(dirichlet_workset);
-
-  auto     col_is_dbc_data = Albany::getLocalData(col_is_dbc_.getConst());
-  auto     range_spmd_vs   = Albany::getSpmdVectorSpace(J->range());
-  const LO num_local_rows  = range_spmd_vs->localSubDim();
+    auto tCol2DBC = Albany::getTpetraVector(col_is_dbc_);
+    auto col2DBCView = tCol2DBC->getLocalViewDevice(Tpetra::Access::ReadOnly);
+    auto col2DBC = Kokkos::subview(col2DBCView, Kokkos::ALL(), 0);
+
+    auto tJac = Albany::getTpetraMatrix(dirichlet_workset.Jac);
+    auto lJac = tJac->getLocalMatrixDevice();
+    auto numLRows = lJac.numRows();
+
+    using range_policy = Kokkos::RangePolicy<PHX::Device::execution_space>;
+    Kokkos::parallel_for(
+      "SDirichletField<Jacobian>::evaluateFields",
+      range_policy(0, numLRows), KOKKOS_LAMBDA(const int lrow) {
+        const bool isRowDBC = col2DBC(lrow) > 0;
+        if (fill_residual == true && isRowDBC)
+          fView(lrow, 0) = 0.0;
+
+        auto lJacRow = lJac.row(lrow);
+        auto numCols = lJacRow.length;
+        for (LO i = 0; i < numCols; ++i) {
+          auto lcol = lJacRow.colidx(i);
+          const bool isDiagEntry = lcol == lrow;
+          if (isDiagEntry) continue;
+
+          const bool isColDBC = col2DBC(lcol) > 0;
+          if (isRowDBC || isColDBC)
+            lJacRow.value(i) = 0.0;
+        }
+      });
+  }
+  else {
+    Teuchos::RCP<Thyra_LinearOp>     J = dirichlet_workset.Jac;
+    auto f_view = fill_residual ? Albany::getNonconstLocalData(f) : Teuchos::null;
+    Teuchos::Array<ST> entries;
+    Teuchos::Array<LO> indices;
+
+    auto     col_is_dbc_data = Albany::getLocalData(col_is_dbc_.getConst());
+    auto     range_spmd_vs   = Albany::getSpmdVectorSpace(J->range());
+    const LO num_local_rows  = range_spmd_vs->localSubDim();
+    const std::vector<std::vector<int> >& nsNodes = dirichlet_workset.nodeSets->find(this->nodeSetID)->second;
 
-  for (LO local_row = 0; local_row < num_local_rows; ++local_row) {
-    Albany::getLocalRowValues(J, local_row, indices, entries);
-    auto row_is_dbc = col_is_dbc_data[local_row] > 0;
-    if (row_is_dbc && fill_residual == true) {
-      int lunk = nsNodes[local_row][this->offset];
-      f_view[lunk] = 0.0;
-    }
+    for (LO local_row = 0; local_row < num_local_rows; ++local_row) {
+      Albany::getLocalRowValues(J, local_row, indices, entries);
+      auto row_is_dbc = col_is_dbc_data[local_row] > 0;
+      if (row_is_dbc && fill_residual == true) {
+        int lunk = nsNodes[local_row][this->offset];
+        f_view[lunk] = 0.0;
```

Comment:
Is this correct? My understanding is that nsNodes is dimensioned by the list of nodes on the boundary, not the number of local matrix rows. I think there's an attempt to grab the local row, `lunk` but isn't `lunk == local_row`? If I can't find a test which actually uses this, I'll try to print everything to make sure my understanding is correct.

Reply:
It seems to me like the number of local rows in the Jacobian matrix would include all components. Also I think if nsNodes is any subset of all nodes in the system, nsNodes[local_row] would go out-of-bounds. I'll try dumping values to confirm.

Reply:
This is only needed in the residual vector which I don't think is changed in this evaluator for ALI (at least I haven't seen it changed - still looking for a case that uses this code). The residual evaluator does the right thing I think.

Reply:
I realize I don't really understand this code. What do you mean with residual vector?

Reply:
The `f` vector.

Reply:
How can it not happen for ALI? In our ME, every call to computeGlobalJacobian/Tangent also passes `f_out`, to try to compute the res alongside the jac/tangent.

Reply:
f and J are passed but when computeGlobalJacobian is called, it seems to only do the J portions when evaluateFields<Jacobian> is called. I think f is filled when computeGlobalResidual is called. It does seem like at some point these were both being computed in computeGlobalJacobian.

Reply:
> f and J are passed but when computeGlobalJacobian is called, it seems to only do the J portions when evaluateFields<Jacobian> is called. I think f is filled when computeGlobalResidual is called. It does seem like at some point these were both being computed in computeGlobalJacobian.

That was my understanding: if f wasn't evaluated along with Jacobian/Tangent/DistParamDeriv, _then_ call computeGlobalResidual (as last resort to get f). I wonder if something changed in NOX/Piro, to cause f to be evaluated before (or at least not together with) other stuff...

Reply:
@jewatkins I came to the conclusion that we should remove that piece of the code:
https://github.com/sandialabs/Albany/blob/dc81ea96bc6946668b72cacf5ccba41a48360afb/src/evaluators/bc/PHAL_SDirichletField_Def.hpp#L176-L199

in fact this is done, and correctly, in the preEvaluate:
https://github.com/sandialabs/Albany/blob/dc81ea96bc6946668b72cacf5ccba41a48360afb/src/evaluators/bc/PHAL_SDirichletField_Def.hpp#L297-L320

I still haven't figured out why f is always null there.

Reply:
Okay will do. I verified the number of local rows is much larger than the number of node set nodes in the test I modified so this would be a bug if it were executed.